File: backend/app/api/v1/reports.py
"""Attendance summary reports with CSV/JSON export (admin only)."""
import csv
import io
import uuid
from datetime import date

from fastapi import APIRouter, Depends, Query
from fastapi.responses import StreamingResponse
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_db
from app.permissions.rbac import require_admin
from app.schemas.dashboard import EmployeeSummaryRow
from app.services.dashboard_service import DashboardService

router = APIRouter()


@router.get(
    "/attendance-summary",
    response_model=list[EmployeeSummaryRow],
    dependencies=[Depends(require_admin)],
)
async def attendance_summary(
    date_from: date,
    date_to: date,
    department_id: uuid.UUID | None = None,
    db: AsyncSession = Depends(get_db),
) -> list[EmployeeSummaryRow]:
    return await DashboardService(db).attendance_summary(date_from, date_to, department_id)


@router.get("/attendance-summary/export", dependencies=[Depends(require_admin)])
async def export_attendance_summary(
    date_from: date,
    date_to: date,
    department_id: uuid.UUID | None = None,
    format: str = Query(default="csv", pattern="^csv$"),
    db: AsyncSession = Depends(get_db),
) -> StreamingResponse:
    rows = await DashboardService(db).attendance_summary(date_from, date_to, department_id)

    buffer = io.StringIO()
    writer = csv.writer(buffer)
    writer.writerow(
        [
            "Employee Code",
            "Full Name",
            "Present Days",
            "Absent Days",
            "Late Days",
            "Total Working Minutes",
            "Total Overtime Minutes",
        ]
    )
    for row in rows:
        writer.writerow(
            [
                row.employee_code,
                row.full_name,
                row.present_days,
                row.absent_days,
                row.late_days,
                row.total_working_minutes,
                row.total_overtime_minutes,
            ]
        )
    buffer.seek(0)

    filename = f"attendance-summary-{date_from}-to-{date_to}.csv"
    return StreamingResponse(
        iter([buffer.getvalue()]),
        media_type="text/csv",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )

File: frontend/src/api/reports.ts
import { apiClient } from "@/api/client";

export interface EmployeeSummaryRow {
  employee_id: string;
  employee_code: string;
  full_name: string;
  present_days: number;
  absent_days: number;
  late_days: number;
  total_working_minutes: number;
  total_overtime_minutes: number;
}

export async function fetchAttendanceSummary(
  dateFrom: string,
  dateTo: string,
): Promise<EmployeeSummaryRow[]> {
  const { data } = await apiClient.get<EmployeeSummaryRow[]>("/reports/attendance-summary", {
    params: { date_from: dateFrom, date_to: dateTo },
  });
  return data;
}

export function attendanceSummaryExportUrl(dateFrom: string, dateTo: string): string {
  const base = apiClient.defaults.baseURL;
  return `${base}/reports/attendance-summary/export?date_from=${dateFrom}&date_to=${dateTo}`;
}
